zcash_client_sqlite/src/wallet/scanning.rs

```diff
@@ -0,0 +1,1107 @@
+use rusqlite::{self, named_params, types::Value, OptionalExtension};
+use std::cmp::{max, min, Ordering};
+use std::collections::BTreeSet;
+use std::ops::{Not, Range};
+use std::rc::Rc;
+use zcash_client_backend::data_api::scanning::{ScanPriority, ScanRange};
+
+use incrementalmerkletree::{Address, Position};
+use zcash_primitives::consensus::{self, BlockHeight, NetworkUpgrade};
+
+use zcash_client_backend::data_api::SAPLING_SHARD_HEIGHT;
+
+use crate::error::SqliteClientError;
+use crate::{PRUNING_DEPTH, VALIDATION_DEPTH};
+
+use super::block_height_extrema;
+
+#[derive(Debug, Clone, Copy)]
+enum Insert {
+    Left,
+    Right,
+}
+
+impl Not for Insert {
+    type Output = Self;
+
+    fn not(self) -> Self::Output {
+        match self {
+            Insert::Left => Insert::Right,
+            Insert::Right => Insert::Left,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum Dominance {
+    Left,
+    Right,
+    Equal,
+}
+
+impl From<Insert> for Dominance {
+    fn from(value: Insert) -> Self {
+        match value {
+            Insert::Left => Dominance::Left,
+            Insert::Right => Dominance::Right,
+        }
+    }
+}
+
+pub(crate) fn parse_priority_code(code: i64) -> Option<ScanPriority> {
+    use ScanPriority::*;
+    match code {
+        10 => Some(Scanned),
+        20 => Some(Historic),
+        30 => Some(OpenAdjacent),
+        40 => Some(FoundNote),
+        50 => Some(ChainTip),
+        60 => Some(Verify),
+        _ => None,
+    }
+}
+
+pub(crate) fn priority_code(priority: &ScanPriority) -> i64 {
+    use ScanPriority::*;
+    match priority {
+        Scanned => 10,
+        Historic => 20,
+        OpenAdjacent => 30,
+        FoundNote => 40,
+        ChainTip => 50,
+        Verify => 60,
+    }
+}
+
+pub(crate) fn suggest_scan_ranges(
+    conn: &rusqlite::Connection,
+    min_priority: Option<ScanPriority>,
```

Comment:
Since this is internal, I would suggest not having a default and passing `ScanPriority::Historic` when needed. (Explicit Is Better Than Implicit.)